### DISCUSSION TITLE
How to load ROM from a file?

### DISCUSSION THREAD

**Discussion by wconrad** [hneemann/Digital#787] [Q&A] [answered]
How can I load a ROM from a file?

# What I've tried:

I can add a ROM to a circuit, and use the EDIT dialog to set its data.  This data is persisted with the circuit just fine: If I close Digital and then reload the circuit, the ROM has the contents I set in the EDIT dialog.

The ROM dialog has "SAVE" and "LOAD" buttons, but they are inactive/grayed-out:

![rom1](https://user-images.githubusercontent.com/1935752/124190826-bf470c80-da77-11eb-9146-1ed8ca9b1b25.png)

# Versions

```
OS: Linux 
OS version: 4.19.0-17-amd64 
JRE: 11.0.11 
VM: 11.0.11+9-post-Debian-1deb10u1 

Digital.jar 
Build git-Revision: v0.27 
Build Time: 2021-04-09 09:17
```

# Why I think this must be possible

Issue #521 references "loading a raw binary file into a ROM," so I think this must be possible.  I just can't figure out how.

# The use case

The use case is that a ROM will hold a CPU's microprogram.  The microprogram will be generated from an external microprogram assembler that I will write.  I will need to get the assembled microprogram from the assembler into the ROM somehow.

**Comment by wconrad** [comment]
# A way to do it

I don't know if this is _the_ way to do it, but it's _a_ way.

I edited the circuit's XML file and added a load path entry to the ROM:

```
       <elementName>ROM</elementName>
      <elementAttributes>
        ...
        <entry>
          <string>lastDataFile</string>
          <string>/home/wayne/lab/Digital/wayne/test.hex</string>
        </entry>
      </elementAttributes>
```

After I reloaded the circuit, I was able to use the "SAVE" button to write out a .hex file, which created a file like this:

```
v2.0 raw
1
2
3
4
```

After editing that file, the RELOAD button worked to load the edited contents into the ROM.

**Comment by TheWildJarvi** [comment, marked as answer]
click edit, then you can save and load from the file header.

**Reply by wconrad** [reply]
Thanks!
